### PR TITLE
Import files to working directory

### PR DESCRIPTION
**Pull request by mstajdohar**

### CODE REVIEW

File: test_resolwe_runtime_utils.py [hunk 1]
```diff
@@ -279,27 +294,154 @@ class TestCheckRC(ResolweRuntimeUtilsTestCase):
...
+                              'test uncompressed.txt',
+                              'test uncompressed.txt')
+
+    def test_gz(self):
+        self.assertImportFile('some file.1.txt.gz',
+                              'test gz.txt.gz',
+                              'test gz.txt')
+
+    def test_7z(self):
+        self.assertImportFile('some file.1.txt.zip',
```
Comment (mzganec) [inline test_resolwe_runtime_utils.py]: It seems that the input file is missing (see the error message on Travis). I get a similar error locally.

File: resolwe_runtime_utils.py [hunk 2]
```diff
@@ -275,12 +288,177 @@ def checkrc(rc, *args):
...
+                        max_progress * total / downloaded, 2)
+
+                    if next_progress > current_progress:
+                        print(progress(current_progress))
+
+                    current_progress = next_progress
+
+        # Check if a temporary file exists.
+        if not os.path.isfile(file_name):
+            error("File transfer failed: temporary file not found {}".format(file_name))
```
Comment (mzganec) [inline resolwe_runtime_utils.py]: This line doesn't do anything. Perhaps a missing `print`?

File: resolwe_runtime_utils.py [hunk 3]
```diff
@@ -275,12 +288,177 @@ def checkrc(rc, *args):
...
+
+        src = file_name
+    else:
+        if not os.path.isfile(src):
+            error("Source file not found {}".format(src))
```
Comment (mzganec) [inline resolwe_runtime_utils.py]: Missing `print`?

File: resolwe_runtime_utils.py [hunk 4]
```diff
@@ -275,12 +288,177 @@ def checkrc(rc, *args):
...
+    """Import file to working directory.
+
+    :param src: Source file path or URL
+    :param file_name: Source file name
+    :param imported_format: Import file format (extracted, compressed or both)
+    :param max_progress: Progress value after import
+    :return: Destination file path (if extracted and compressed, extracted path given)
+    """
+    print("Importing and compressing {}...".format(file_name))
+
```
Comment (mzganec) [inline resolwe_runtime_utils.py]: Given how some tests fail now, consider adding a check prior to extraction that the input file actually exists. It'll be easier to debug if we have a dedicated error message so that we don't need to dig into the output of `subprocess.check_call`.

File: resolwe_runtime_utils.py [hunk 5]
```diff
@@ -275,12 +288,177 @@ def checkrc(rc, *args):
...
+        The file_name must have .gz extension.
+        """
+        if imported_format != ImportedFormat.EXTRACTED:  # Compressed file required
+            with gzip.open(src, 'rb') as f:
+                try:
+                    while f.read(CHUNK_SIZE) != b'':
+                        pass
+                except Exception:
+                    raise ValueError("invalid gzip file format")
```
Comment (mzganec) [inline resolwe_runtime_utils.py]: Consider including the file name. Don't we want to `print(error(...))` here as well?

File: resolwe_runtime_utils.py [hunk 6]
```diff
@@ -275,12 +288,177 @@ def checkrc(rc, *args):
...
+            total = float(total) if total else None
+            downloaded = 0
+            current_progress = 0
+            for content in response.iter_content(chunk_size=CHUNK_SIZE):
+                f.write(content)
+
+                if total is not None and max_progress is not None:
+                    downloaded += len(content)
+                    next_progress = round(
+                        max_progress * total / downloaded, 2)
```
Comment (mzganec) [inline resolwe_runtime_utils.py]: This is going to work a bit funny because we don't have `min_progress`. Consider the case where the progress just before running `import_file` is not 0, e.g. if we import multiple files or do some processing prior to running `import_file`.

File: test_resolwe_runtime_utils.py (hunk 1 above)
Comment (mstajdohar) [reply test_resolwe_runtime_utils.py]: Files added

File: resolwe_runtime_utils.py (hunk 2 above)
Comment (mstajdohar) [reply resolwe_runtime_utils.py]: Added

File: resolwe_runtime_utils.py (hunk 3 above)
Comment (mstajdohar) [reply resolwe_runtime_utils.py]: Added

File: resolwe_runtime_utils.py (hunk 5 above)
Comment (mstajdohar) [reply resolwe_runtime_utils.py]: Refactored to print

File: resolwe_runtime_utils.py (hunk 6 above)
Comment (mstajdohar) [reply resolwe_runtime_utils.py]: Changed to `import_from`, `import_to`. Please review.

File: resolwe_runtime_utils.py [hunk 7]
```diff
@@ -275,14 +301,224 @@ def checkrc(rc, *args):
...
+            raise ValueError("progress_to must be higher than progress_from")
+
+    print("Importing and compressing {}...".format(file_name))
+
+    def importGz():
+        """Import gzipped file.
+
+        The file_name must have .gz extension.
+        """
+        if imported_format != ImportedFormat.EXTRACTED:  # Compressed file required
```
Comment (mzganec) [inline resolwe_runtime_utils.py]: I suggest to use a different order of operations here. If the extracted file is requested, first extract the file to avoid reading the compressed file twice and simultaneously verify compressed file integrity in a `try` .. `except` block. If the compressed file is needed, you can just [move it instead of copying](https://github.com/genialis/resolwe/blob/master/resolwe/toolkit/docker_images/base/re-import.sh#L50).

File: resolwe_runtime_utils.py [hunk 8]
```diff
@@ -275,14 +301,224 @@ def checkrc(rc, *args):
...
+            with gzip.open(src, 'rb') as f:
+                try:
+                    while f.read(CHUNK_SIZE) != b'':
+                        pass
+                except Exception:
+                    print(error("Invalid gzip file format"))
```
Comment (mzganec) [inline resolwe_runtime_utils.py]: Consider adding `src` to the error message.

File: resolwe_runtime_utils.py [hunk 9]
```diff
@@ -275,14 +301,224 @@ def checkrc(rc, *args):
...
+            else file_name[:-3]
+        )
+
+    def import7z():
+        """Import compressed file in various formats.
+
+        Supported extensions: .bz2, .zip, .rar, .7z, .tar.gz, and .tar.bz2.
+        """
+        extracted_name, _ = os.path.splitext(file_name)
+        temp_dir = 'temp_{}'.format(extracted_name)
```
Comment (mzganec) [inline resolwe_runtime_utils.py:396]: A non-empty `temp_dir` can already exist prior to extraction. Consider using a [temporary directory](https://docs.python.org/3.7/library/tempfile.html#tempfile.TemporaryDirectory) instead.

File: resolwe_runtime_utils.py [hunk 10]
```diff
@@ -275,14 +301,224 @@ def checkrc(rc, *args):
...
+
+    def importUncompressed():
+        """Import uncompressed file."""
+        if imported_format != ImportedFormat.EXTRACTED:  # Compressed file required
+            with open(src, 'rb') as f_in, gzip.open(file_name + '.gz', 'wb') as f_out:
+                shutil.copyfileobj(f_in, f_out, CHUNK_SIZE)
+
+        if imported_format != ImportedFormat.COMPRESSED:  # Extracted file required
+            try:
+                shutil.copyfile(src, file_name)
```
Comment (mzganec) [inline resolwe_runtime_utils.py:453]: Move.

File: resolwe_runtime_utils.py [hunk 11]
```diff
@@ -275,14 +301,224 @@ def checkrc(rc, *args):
...
+            total = float(total) if total else None
+            downloaded = 0
+            current_progress = 0
+            for content in response.iter_content(chunk_size=CHUNK_SIZE):
+                f.write(content)
+
+                if total is not None and progress_to is not None:
+                    downloaded += len(content)
+                    progress_span = progress_to - progress_from
+                    next_progress = progress_from + progress_span * total / downloaded
```
Comment (mzganec) [inline resolwe_runtime_utils.py]: We need the inverse here, `downloaded / total`. Can `total` be 0?

File: resolwe_runtime_utils.py [hunk 12]
```diff
@@ -275,14 +301,224 @@ def checkrc(rc, *args):
...
+                    next_progress = round(next_progress, 2)
+
+                    if next_progress > current_progress:
+                        print(progress(current_progress))
```
Comment (mzganec) [inline resolwe_runtime_utils.py]: Replace `current_progress` with `next_progress`, otherwise the first progress update will set it to 0. Note that even with this change the progress may never be set to `progress_from` but I don't consider this to be an issue.

File: resolwe_runtime_utils.py [hunk 13]
```diff
@@ -275,14 +301,224 @@ def checkrc(rc, *args):
...
+                break
+
+        if token is not None:
+            params = {'confirm': token}
+            response = session.get(src, params=params, stream=True)
+
+    elif re.match(
+        r'^(https?|ftp)://[-A-Za-z0-9\+&@#/%?=~_|!:,.;]*[-A-Za-z0-9\+&@#/%=~_|]$', src
+    ):
+        response = requests.get(src, stream=True)
```
Comment (mzganec) [inline resolwe_runtime_utils.py]: Uncaught exception.

File: resolwe_runtime_utils.py [hunk 14]
```diff
@@ -275,14 +301,224 @@ def checkrc(rc, *args):
...
+                        print(progress(current_progress))
+                        current_progress = next_progress
+
+        # Check if a temporary file exists.
+        if not os.path.isfile(file_name):
+            print(
+                error("Download failed: temporary file not found {}".format(file_name))
+            )
+
+        src = file_name
```
Comment (mzganec) [inline resolwe_runtime_utils.py:514]: Unused assignment.

File: resolwe_runtime_utils.py [hunk 15]
```diff
@@ -275,14 +301,224 @@ def checkrc(rc, *args):
...
+    src,
+    file_name,
+    imported_format=ImportedFormat.BOTH,
+    progress_from=0.0,
+    progress_to=None,
+):
+    """Import file to working directory.
+
+    :param src: Source file path or URL
+    :param file_name: Source file name
```
Comment (mzganec) [inline resolwe_runtime_utils.py:335]: Destination file name?

File: resolwe_runtime_utils.py [hunk 16]
```diff
@@ -275,14 +301,224 @@ def checkrc(rc, *args):
...
+
+        src = file_name
+    else:
+        if not os.path.isfile(src):
+            print(error("Source file not found {}".format(src)))
+
+    # Decide which import should be used.
+    if re.search(r'\.(bz2|zip|rar|7z|tgz|tar\.gz|tar\.bz2)$', file_name):
+        destination_file_name = import7z()
+    elif file_name.endswith('.gz'):
```
Comment (mzganec) [inline resolwe_runtime_utils.py:522]: `7z` handles `.gz` files as well. So special treatment of `.gz` files is redundant unless we need to handle `.gz` files even if `7z` is not installed.

File: resolwe_runtime_utils.py [hunk 17]
```diff
@@ -275,14 +301,224 @@ def checkrc(rc, *args):
...
+        destination_file_name = importGz()
+    else:
+        destination_file_name = importUncompressed()
+
+    if progress_to is not None:
+        print(progress(progress_to))
+
+    return destination_file_name
```
Comment (mzganec) [inline resolwe_runtime_utils.py:530]: The case of overwriting an existing file named `destination_file_name` is unhandled. This can occur in processes with multiple `basic:file` inputs where the user selects local files with the same file name in different directories. In processes we normally keep the same file name as on user's disk.

Depending on the process, the data object can finish computing without errors, but the results will be wrong. We haven't noticed such cases yet though.

File: resolwe_runtime_utils.py [hunk 18]
```diff
@@ -275,14 +301,224 @@ def checkrc(rc, *args):
...
+        )
+
+    # Large file download from Google Drive requires cookie and token.
+    response = None
+    if re.match(
+        r'^https://drive.google.com/[-A-Za-z0-9\+&@#/%?=~_|!:,.;]*[-A-Za-z0-9\+&@#/%=~_|]$',
+        src,
+    ):
+        session = requests.Session()
+        response = session.get(src, stream=True)
```
Comment (mzganec) [inline resolwe_runtime_utils.py]: Uncaught exception.

File: resolwe_runtime_utils.py (hunk 7 above)
Comment (mstajdohar) [reply resolwe_runtime_utils.py]: I have changed the order as suggested, thanks. Moving the file does not give us much advantage as most of our deploys have the upload folder on a different physical hard drive. Furthermore, we plan to remove files from the upload folder in `purge` if the process completed successfully. This will allow for data recovery in case of errors.

File: resolwe_runtime_utils.py (hunk 8 above)
Comment (mstajdohar) [reply resolwe_runtime_utils.py]: Thanks. I have decided to report `file_name` as src is usually a temporary name (UUID) assigned during upload.

File: resolwe_runtime_utils.py (hunk 9 above)
Comment (mstajdohar) [reply resolwe_runtime_utils.py:396]: This is highly unlikely because the name is `temp_<file_name>`. Also, the temporary directory is removed after import so the next call to re-import can not cause this error. 

We have a greater problem when users would upload files of the same name which is problematic in all use cases. However, there are no known occurrences of this issue.

File: resolwe_runtime_utils.py (hunk 10 above)
Comment (mstajdohar) [reply resolwe_runtime_utils.py:453]: Nope. We want to ensure repeatability in case of errors.

File: resolwe_runtime_utils.py (hunk 11 above)
Comment (mstajdohar) [reply resolwe_runtime_utils.py]: Thanks.

> Can total be 0?

Probably not. It's either None or a positive integer. But you reminded me I should check that `progress_to` and `progress_from` values are between 0 and 1.

File: resolwe_runtime_utils.py (hunk 14 above)
Comment (mstajdohar) [reply resolwe_runtime_utils.py:514]: Functions `importGz`, `import7z` and `importUncompressed` use `src`.

It may be ugly but not unused ;)

File: resolwe_runtime_utils.py (hunk 15 above)
Comment (mstajdohar) [reply resolwe_runtime_utils.py:335]: Well, this is actually the name of the input file. Coincidently this is also the destination name.

Here's an example. If I would upload `my_reads.fq.gz`, `src` would be `e56d8bdc-56ad-11e9-8647-d663bd873d93` and `file_name` would be `my_reads.fq.gz`.

File: resolwe_runtime_utils.py (hunk 16 above)
Comment (mstajdohar) [reply resolwe_runtime_utils.py:522]: Gz is handled separately, because we can simply copy compressed file. Because it is handled separately and because we know the file is gzipped, it makes the most sense to use Python's gzip implementation.

File: resolwe_runtime_utils.py (hunk 17 above)
Comment (mstajdohar) [reply resolwe_runtime_utils.py:530]: True. This is problematic. Python process will crash if this happens, but most likely with an ugly error.

Since there are no known occurrences of this issue, we may want to fix it in the future.

File: resolwe_runtime_utils.py (hunk 18 above)
Comment (mstajdohar) [reply resolwe_runtime_utils.py]: I've decided to reverse the logic and change `print(error(...))` with raise exception. I will update Python processes to catch exceptions and automatically save process errors.

File: .travis.yml [hunk 19]
```diff
@@ -27,15 +27,42 @@ jobs:
     - stage: test
       env: TOX_ENV=py34
       python: "3.4"
+      before_install: skip
```
Comment (dblenkus) [inline .travis.yml]: This is not needed as there is no `before_install` to override.

File: docs/contributing.rst [hunk 20]
```diff
@@ -106,29 +100,41 @@ Commit changes to git::
...
-    tox -r
+    git push --set-upstream <fork-name> release-<new-version>
+
+Wait for the tests to pass and the pull request to be approved. Merge the code
+to master::
 
-Create source distribution::
+    git checkout master
+    git merge --ff-only release-<new-version>
+    git push <upstream-name> master <new-version>
```
Comment (dblenkus) [inline docs/contributing.rst]: `<new-version>` is only tagged in the next step, you can probably omit it here.

File: .travis.yml [hunk 21]
```diff
@@ -27,15 +27,42 @@ jobs:
...
       env: TOX_ENV=docs
+      before_install: skip
     - stage: test
       env: TOX_ENV=packaging
+      before_install: skip
+
+    # NOTE: We undo/change all the global Travis CI step definitions to ensure
+    # building of source and wheel distributions is performed in a clean
+    # environment.
+    - stage: release
```
Comment (dblenkus) [inline .travis.yml:51]: you need to override `after_success` here. Check [this](https://github.com/genialis/resolwe/blob/81a3e0854a8b531f8bdda1ae34658defd7598ded/.travis.yml#L116) for details.

File: .travis.yml [hunk 22]
```diff
@@ -27,23 +27,22 @@ jobs:
     - stage: test
       env: TOX_ENV=py34
       python: "3.4"
-      before_install: skip
```
Comment (dblenkus) [inline .travis.yml]: This shouldn't be a part of this commit.

File: tox.ini [hunk 23]
```diff
@@ -1,32 +1,45 @@
...
+    pip check
+
 commands =
-# install testing requirements
-    pip install .[test]
-# run tests
-    py.test --verbose --cov=resolwe_runtime_utils
+    # General tests commands:
+    # Run tests.
+    py34: pytest --verbose --cov=resolwe_runtime_utils
```
Comment (dblenkus) [inline tox.ini]: You can probably write this in a single line using `py3{4,5,6,7}`

File: MANIFEST.in [hunk 24]
```diff
@@ -7,3 +7,4 @@ include docs/conf.py
 include AUTHORS LICENSE
 # Include files needed by tests
 include pyproject.toml
+recursive-include files *
```
Comment (dblenkus) [inline MANIFEST.in]: Rename this to `test_files`.

File: test_resolwe_runtime_utils.py [hunk 25]
```diff
@@ -15,19 +15,48 @@
...
+import responses
+import requests
+
 from resolwe_runtime_utils import (
-    save, export, save_list, save_file, save_file_list, save_dir, save_dir_list,
-    info, warning, error, progress, checkrc,
-    _re_save_main, _re_export_main, _re_save_list_main, _re_save_file_main, _re_save_file_list_main,
-    _re_save_dir_main, _re_save_dir_list_main, _re_info_main, _re_warning_main, _re_error_main,
-    _re_progress_main, _re_checkrc_main
+    save,
```
Comment (dblenkus) [inline test_resolwe_runtime_utils.py:28]: This should probably be part of "black" commit.

**Review by dblenkus — COMMENTED** [review]
You can remove the first commit as you later introduce black.

File: resolwe_runtime_utils.py [hunk 26]
```diff
@@ -275,14 +301,233 @@ def checkrc(rc, *args):
...
+    print("Importing and compressing {}...".format(file_name))
+
+    def importGz():
+        """Import gzipped file.
+
+        The file_name must have .gz extension.
+        """
+        if imported_format != ImportedFormat.COMPRESSED:  # Extracted file required
+            with open(file_name[:-3], 'wb') as f_out, gzip.open(src, 'rb') as f_in:
+                shutil.copyfileobj(f_in, f_out, CHUNK_SIZE)
```
Comment (mzganec) [inline resolwe_runtime_utils.py]: Uncaught exception. [`shutil.copyfileobj` uses `f.read` internally](https://github.com/python/cpython/blob/3.7/Lib/shutil.py#L79) so this line needs to be in the `try` ... `except` block as well.

File: resolwe_runtime_utils.py [hunk 27]
```diff
@@ -275,14 +301,233 @@ def checkrc(rc, *args):
...
+
+    def import7z():
+        """Import compressed file in various formats.
+
+        Supported extensions: .bz2, .zip, .rar, .7z, .tar.gz, and .tar.bz2.
+        """
+        extracted_name, _ = os.path.splitext(file_name)
+        temp_dir = 'temp_{}'.format(extracted_name)
+
+        cmd = '7z x -y -o{} {}'.format(shlex.quote(temp_dir), shlex.quote(src))
```
Comment (mzganec) [inline resolwe_runtime_utils.py:398]: We extract and re-compress `tar.gz` files which could be just copied.

If copying is the only reason why we handle `gz` (without `tar`) files separately, reconsider handling `gz` files here as well. Besides simplifying the code we get parallel decompression for free.

File: resolwe_runtime_utils.py [hunk 28]
```diff
@@ -275,14 +301,233 @@ def checkrc(rc, *args):
...
+
+            if imported_format != ImportedFormat.COMPRESSED:  # Extracted file required
+                shutil.move(temp_file, './{}'.format(extracted_name))
+
+                if extracted_name.endswith('.tar'):
+                    with tarfile.open(extracted_name) as tar:
+                        tar.extractall()
+
+                    os.remove(extracted_name)
+                    extracted_name, _ = os.path.splitext(extracted_name)
```
Comment (mzganec) [inline resolwe_runtime_utils.py]: `extracted_name` can cause problems here. Run `import_file` with a source file `file_name = "src.tar.gz"` which contains a directory named `abc/`. After extraction you are left with the extracted directory `abc/` but `import_file` returns `"src"`. Note that there is no file or directory named `src` after extraction.

File: resolwe_runtime_utils.py [hunk 29]
```diff
@@ -275,14 +301,233 @@ def checkrc(rc, *args):
...
+                    for fname in glob.glob(os.path.join(temp_dir, '*')):
+                        tar.add(fname, os.path.basename(fname))
+
+            if imported_format != ImportedFormat.COMPRESSED:  # Extracted file required
+                for path in os.listdir(temp_dir):
+                    shutil.move(os.path.join(temp_dir, path), './{}'.format(path))
+
+        shutil.rmtree(temp_dir)
+        return (
+            extracted_name + '.gz'
```
Comment (mzganec) [inline resolwe_runtime_utils.py]: If the archive contains one or more directories or several files (this case is handled in the [`else` block](https://github.com/genialis/resolwe-runtime-utils/pull/6/files#diff-1815e39b348d47412791b8389c9ed432R415) above), the value returned here is incorrect. For example, we import a zipped directory `file_name = "src.zip"` and request compressed output. The produced output file will be named `src.tar.gz` but `import_file` returns `"src.gz"`.

Given that all tests pass, perhaps they should be more thorough. :wink:

File: resolwe_runtime_utils.py [hunk 30]
```diff
@@ -275,14 +301,233 @@ def checkrc(rc, *args):
...
+    def import7z():
+        """Import compressed file in various formats.
+
+        Supported extensions: .bz2, .zip, .rar, .7z, .tar.gz, and .tar.bz2.
+        """
+        extracted_name, _ = os.path.splitext(file_name)
+        temp_dir = 'temp_{}'.format(extracted_name)
+
+        cmd = '7z x -y -o{} {}'.format(shlex.quote(temp_dir), shlex.quote(src))
+        subprocess.check_call(cmd, shell=True)
```
Comment (mzganec) [inline resolwe_runtime_utils.py]: Unhandled `subprocess.CalledProcessError`.

File: resolwe_runtime_utils.py [hunk 31]
```diff
@@ -275,14 +301,233 @@ def checkrc(rc, *args):
...
+            src,
+        ):
+            response = requests.get(src, stream=True)
+    except requests.exceptions.ConnectionError:
+        raise requests.exceptions.ConnectionError("could not connect to {}".format(src))
+
+    if response:
+        with open(file_name, 'wb') as f:
+            total = response.headers.get('content-length')
+            total = float(total) if total else None
```
Comment (mzganec) [inline resolwe_runtime_utils.py:494]: `total` is not cast to `int` for compatibility with python2 in [computing progress update](https://github.com/genialis/resolwe-runtime-utils/pull/6/files#diff-1815e39b348d47412791b8389c9ed432R491)?

File: resolwe_runtime_utils.py [hunk 32]
```diff
@@ -275,14 +301,233 @@ def checkrc(rc, *args):
...
+            downloaded = 0
+            current_progress = 0
+            for content in response.iter_content(chunk_size=CHUNK_SIZE):
+                f.write(content)
```
Comment (mzganec) [inline resolwe_runtime_utils.py:498]: After download the original file is not deleted, even if the file is not requested by `imported_format`.

File: resolwe_runtime_utils.py [hunk 33]
```diff
@@ -275,14 +301,233 @@ def checkrc(rc, *args):
...
+            r'^(https?|ftp)://[-A-Za-z0-9\+&@#/%?=~_|!:,.;]*[-A-Za-z0-9\+&@#/%=~_|]$',
+            src,
+        ):
+            response = requests.get(src, stream=True)
+    except requests.exceptions.ConnectionError:
+        raise requests.exceptions.ConnectionError("could not connect to {}".format(src))
+
+    if response:
+        with open(file_name, 'wb') as f:
+            total = response.headers.get('content-length')
```
Comment (mzganec) [inline resolwe_runtime_utils.py:493]: Just a note, nothing to do here: responses from Google Drive don't return content length so we are left without progress update.

File: resolwe_runtime_utils.py [hunk 34]
```diff
@@ -275,14 +301,233 @@ def checkrc(rc, *args):
...
+            if token is not None:
+                params = {'confirm': token}
+                response = session.get(src, params=params, stream=True)
+
+        elif re.match(
+            r'^(https?|ftp)://[-A-Za-z0-9\+&@#/%?=~_|!:,.;]*[-A-Za-z0-9\+&@#/%=~_|]$',
+            src,
+        ):
+            response = requests.get(src, stream=True)
+    except requests.exceptions.ConnectionError:
```
Comment (mzganec) [inline resolwe_runtime_utils.py:488]: [Errors and Exceptions](http://docs.python-requests.org/en/master/user/quickstart/#errors-and-exceptions)

File: resolwe_runtime_utils.py [hunk 35]
```diff
@@ -275,14 +301,233 @@ def checkrc(rc, *args):
...
+    EXTRACTED = 'extracted'
+    COMPRESSED = 'compressed'
+    BOTH = 'both'
+
+
+def import_file(
+    src,
+    file_name,
+    imported_format=ImportedFormat.BOTH,
+    progress_from=0.0,
```
Comment (mzganec) [inline resolwe_runtime_utils.py:329]: Consider adding here an optional `output_dir` argument, which specifies the destination directory of import. This argument would (1) help prevent overwriting existing files during import. In the case where the source archive contains multiple files or directories, (2) it's easier to see what was imported and (3) `import_file` can then provide a more sensible return value.

**Comment by mstajdohar** [comment]
> You can remove the first commit as you later introduce black.

I could, but this would most likely cause quite a lot of conflicts...

File: .travis.yml (hunk 19 above)
Comment (mstajdohar) [reply .travis.yml]: done

File: docs/contributing.rst (hunk 20 above)
Comment (mstajdohar) [reply docs/contributing.rst]: done

File: .travis.yml (hunk 21 above)
Comment (mstajdohar) [reply .travis.yml:51]: done

File: tox.ini (hunk 23 above)
Comment (mstajdohar) [reply tox.ini]: done. testing...

File: .travis.yml (hunk 22 above)
Comment (mstajdohar) [reply .travis.yml]: done

File: MANIFEST.in (hunk 24 above)
Comment (mstajdohar) [reply MANIFEST.in]: done

File: test_resolwe_runtime_utils.py (hunk 25 above)
Comment (mstajdohar) [reply test_resolwe_runtime_utils.py:28]: done

File: resolwe_runtime_utils.py (hunk 26 above)
Comment (mstajdohar) [reply resolwe_runtime_utils.py]: How should I handle the exception?

Comment (mzganec) [reply resolwe_runtime_utils.py]: I suggest to handle it the same way as 5 lines below.

File: resolwe_runtime_utils.py (hunk 28 above)
Comment (mstajdohar) [reply resolwe_runtime_utils.py]: There is no way around this. If you allow upload of compressed collections (of files and folder), you better be sure you can handle results.

**Review by dblenkus — APPROVED** [review]

File: resolwe_runtime_utils.py (hunk 30 above)
Comment (mstajdohar) [reply resolwe_runtime_utils.py]: How should I handle the exception?

File: resolwe_runtime_utils.py (hunk 31 above)
Comment (mstajdohar) [reply resolwe_runtime_utils.py:494]: We do not support Python2.

File: resolwe_runtime_utils.py (hunk 32 above)
Comment (mstajdohar) [reply resolwe_runtime_utils.py:498]: Cleaned by purge (the same as is the file would be in the upload folder).

File: resolwe_runtime_utils.py (hunk 34 above)
Comment (mstajdohar) [reply resolwe_runtime_utils.py:488]: What do you suggest we do here? The only reason I catch ConnectionError is to point out this is most likely due to invalid URL. Other exceptions are self-explanatory.

File: resolwe_runtime_utils.py (hunk 35 above)
Comment (mstajdohar) [reply resolwe_runtime_utils.py:329]: We have never encountered this use case, so I am not fond of complicating usage. If this is a problem one can always `cd` to a subdir and call `import_file` there. We can add this later if required.

File: resolwe_runtime_utils.py (hunk 30 above)
Comment (mzganec) [reply resolwe_runtime_utils.py]: I would print an error message that something went wrong during decompression. We want to fail gracefully in case the file is corrupted. I suggest to do something similar as in the gzip integrity check above.

File: resolwe_runtime_utils.py (hunk 26 above)
Comment (mstajdohar) [reply resolwe_runtime_utils.py]: Hm. That handles the case where the file name is the the same as downloaded file and we can skip the copy. This can't happen above, because the function handles uncompressed files and here we compress it (so src and file_name) are always different.

File: resolwe_runtime_utils.py (hunk 30 above)
Comment (mstajdohar) [reply resolwe_runtime_utils.py]: I have decided not to print errors in python processes. We should throw exceptions. The Python process engine should catch the exception and reformat it to process message.

If I print an error (re-error) the python code will continue and likely brake at some lower point. This is then even harder to debug.

Comment (mstajdohar) [reply resolwe_runtime_utils.py]: OK, I'll handle the exception

Comment (mstajdohar) [reply resolwe_runtime_utils.py]: done

File: resolwe_runtime_utils.py (hunk 26 above)
Comment (mzganec) [reply resolwe_runtime_utils.py]: We are looking at different sections of the code and I mean [this line](https://github.com/genialis/resolwe-runtime-utils/pull/6/files#diff-1815e39b348d47412791b8389c9ed432R368). We extract compressed `gz` file and raise `ValueError` in the linked line but we don't do that here.

Comment (mstajdohar) [reply resolwe_runtime_utils.py]: Done

File: resolwe_runtime_utils.py (hunk 29 above)
Comment (mstajdohar) [reply resolwe_runtime_utils.py]: fixed (test added)

File: resolwe_runtime_utils.py (hunk 33 above)
Comment (mstajdohar) [reply resolwe_runtime_utils.py:493]: that's annoying

Comment (mstajdohar) [reply resolwe_runtime_utils.py:493]: Do they return something else? Does the browser show progress for responses from Google Drive?

**Comment by mstajdohar** [comment]
@dblenkus 

>> You can remove the first commit as you later introduce black.

> I could, but this would most likely cause quite a lot of conflicts...

Done.

File: resolwe_runtime_utils.py (hunk 33 above)
Comment (mzganec) [reply resolwe_runtime_utils.py:493]: Let me rephrase that. Responses for download of **large files** from Google Drive don't return content length. Large files are the ones that require confirming the download warning. In this case even the browser doesn't display download progress and the following headers are returned:
```
{
    'X-GUploader-UploadID':'AEnB2Ur7WZU4C0NuP7mtzw0mRVmdxm54cL2ksZE_35riixQNBk0bUJh9_6PZbWYbd2OB7NmDEZyH2Ku3oybkRdpmtZ7uU6MKQw',
    'Access-Control-Allow-Origin':'*',
    'Access-Control-Allow-Credentials':'false',
    'Access-Control-Allow-Headers':'Accept, Accept-Language, Authorization, Cache-Control, Content-Disposition, Content-Encoding, Content-Language, Content-Length, Content-MD5, Content-Range, Content-Type, Date, GData-Version, google-cloud-resource-prefix, Host, If-Match, If-Modified-Since, If-None-Match, If-Unmodified-Since, Origin, OriginToken, Pragma, Range, Slug, Transfer-Encoding, Want-Digest, x-chrome-connected, X-ClientDetails, X-Client-Version, X-Firebase-Locale, X-GData-Client, X-GData-Key, X-Goog-AdX-Buyer-Impersonation, X-Goog-Api-Client, X-Goog-AuthUser, X-Goog-PageId, X-Goog-Encode-Response-If-Executable, X-Goog-Correlation-Id, X-Goog-Request-Info, X-Goog-Request-Reason, X-Goog-Experiments, x-goog-iam-authority-selector, x-goog-iam-authorization-token, X-Goog-Spatula, X-Goog-Upload-Command, X-Goog-Upload-Content-Disposition, X-Goog-Upload-Content-Length, X-Goog-Upload-Content-Type, X-Goog-Upload-File-Name, X-Goog-Upload-Header-Content-Length, X-Goog-Upload-Offset, X-Goog-Upload-Protocol, X-Goog-Visitor-Id, X-Goog-FieldMask, X-HTTP-Method-Override, X-JavaScript-User-Agent, X-Pan-Versionid, X-Proxied-User-IP, X-Origin, X-Referer, X-Requested-With, X-Upload-Content-Length, X-Upload-Content-Type, X-Use-HTTP-Status-Code-Override, X-Ios-Bundle-Identifier, X-Android-Package, X-Ariane-Xsrf-Token, X-YouTube-VVT, X-YouTube-Page-CL, X-YouTube-Page-Timestamp, X-Goog-Meeting-Botguardid, X-Goog-Meeting-Debugid, X-Goog-Meeting-Token, X-Client-Data, X-Sfdc-Authorization, MIME-Version, Content-Transfer-Encoding',
    'Access-Control-Allow-Methods':'GET,OPTIONS',
    'Content-Type':'application/x-rpm',
    'Content-Disposition':'attachment;filename="atom.x86_64.rpm";filename*=UTF-8\'\'atom.x86_64.rpm',
    'Date':'Tue, 09 Apr 2019 10:07:44 GMT',
    'Expires':'Tue, 09 Apr 2019 10:07:44 GMT',
    'Cache-Control':'private, max-age=0',
    'X-Goog-Hash':'crc32c=QNomWA==',
    'Transfer-Encoding':'chunked',
    'Server':'UploadServer',
    'Alt-Svc':'quic=":443"; ma=2592000; v="46,44,43,39"'
}
```
Reponses for download of **small files** from Google Drive do have `Content-Length` header and download progress updates normally.

Comment (mstajdohar) [reply resolwe_runtime_utils.py:493]: > In this case even the browser doesn't display download progress

That's annoying.

**Review by mzganec — APPROVED** [review]
Well done!
